src/Umbraco.Core/IO/ShadowFileSystem.cs

```diff
@@ -201,21 +201,21 @@ public void AddFile(string path, Stream stream, bool overrideIfExists)
 
         public IEnumerable<string> GetFiles(string path)
         {
+            return GetFiles(path, null);
+        }
+
+        public IEnumerable<string> GetFiles(string path, string filter)
+        {
             var normPath = NormPath(path);
             var shadows = Nodes.Where(kvp => IsChild(normPath, kvp.Key)).ToArray();
-            var files = _fs.GetFiles(path);
+            var files = filter != null ? _fs.GetFiles(path, filter) : _fs.GetFiles(path);
             return files
                 .Except(shadows.Where(kvp => (kvp.Value.IsFile && kvp.Value.IsDelete) || kvp.Value.IsDir)
                     .Select(kvp => kvp.Key))
                 .Union(shadows.Where(kvp => kvp.Value.IsFile && kvp.Value.IsExist).Select(kvp => kvp.Key))
```

Comment:
I think there's one thing missing here, you are not filtering the files added by the Union statement, which are files that are not in `_fs` (the actual FS) yet but are in the temp FS.